### PR TITLE
Pass through cache information to the guest (leaf 0x80000006)

### PR DESCRIPTION
**Pull request by kalyazin**
## Changes

Pass through the leaf 0x80000006 from the host to the guest.

## Reason

Cache information may be used by the guest software (eg glibc) to decide on a correct implementation of an algorithm.
Newer kernels (5.8+) pass through the 0x80000006 leaf from the host to the guest, but older do not.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

~~- [ ] If a specific issue led to this PR, this PR closes the issue.~~
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
~~- [ ] API changes follow the [Runbook for Firecracker API changes][2].~~
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
~~- [ ] New `TODO`s link to an issue.~~
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: docs/cpu_templates/cpuid-normalization.md [hunk 1]
```diff
@@ -11,19 +11,20 @@ See also: [boot protocol settings](boot-protocol.md)
...
+| Disable PDCM (Perfmon and Debug Capability)                                          | 0x1        | -       | ECX      | 15    |
+| Enable TSC_DEADLINE                                                                  | 0x1        | -       | ECX      | 24    |
+| Enable HYPERVISOR                                                                    | 0x1        | -       | ECX      | 31    |
+| Set HTT value if the microVM's CPU count is greater than 1                           | 0x1        | -       | EDX      | 28    |
+| Insert leaf 0xb, subleaf 0x1 filled with `0` if it is not already present            | 0xb        | 0x1     | all      | all   |
+| Update extended topology enumeration                                                 | 0xb        | all     | EAX      | 4:0   |
+| Update extended topology enumeration                                                 | 0xb        | all     | EBX      | 15:0  |
+| Update extended topology enumeration                                                 | 0xb        | all     | ECX      | 15:8  |
+| Pass through cache information from host                                             | 0x8000000d | -       | all      | all   |
```
Comment (zulinx86) [inline docs/cpu_templates/cpuid-normalization.md:27]: Isn't this 0x80000006 here?

**Review by pb8o — APPROVED** [review]

**Comment by zulinx86** [comment]
Let me post the link of kernel side commit here for future reference.
https://github.com/amazonlinux/linux/commit/43d05de2bee7561b7958628b7c27c89b779de990
https://lore.kernel.org/all/20200415025105.GE12547@linux.intel.com/

File: src/vmm/src/cpu_config/x86_64/cpuid/normalize.rs [hunk 2]
```diff
@@ -424,6 +437,15 @@ impl super::Cpuid {
 
         Ok(())
     }
+
+    // Update extended cache features entry
+    fn update_extended_cache_features(&mut self) -> Result<(), ExtendedCacheFeaturesError> {
+        let guest_leaf_0x80000006 = self
+            .get_mut(&CpuidKey::leaf(0x80000006))
+            .ok_or(ExtendedCacheFeaturesError::MissingLeaf0x80000006)?;
+        guest_leaf_0x80000006.result = cpuid(0x80000006).into();
```
Comment (zulinx86) [inline src/vmm/src/cpu_config/x86_64/cpuid/normalize.rs:446]: Should we also apply this fix patch like what the kernel does?
https://github.com/amazonlinux/linux/commit/eeb69eab57c6604ac90b3fd8e5ac43f24a5535b1
https://lore.kernel.org/all/20220929225203.2234702-2-jmattson@google.com/

**Comment by kalyazin** [comment]
> I'm not so familiar with CPU cache, but let me ask just in case whether the leaf 0x80000006 is consistent with other leaves.
> 
> For example
> 
> * leaf 4 on intel: https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/x86_64/cpuid/intel/normalize.rs
> * leaf 0x80000001{d,e}: https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/x86_64/cpuid/amd/normalize.rs

As far as I know, those modifications are needed to set the number of cores sharing the cache to match the actual VM size. I don't think they change other cache properties and make them incompatible with the 0x80000006.

**Review by pb8o — APPROVED** [review]

**Pull request merged**